### DISCUSSION TITLE
Test and utilities improvements

### DISCUSSION THREAD

**Discussion by EchterAgo** [openzfsonwindows/openzfs#291] [Ideas]
I'm starting this discussion for ideas about the tests. There are some things that can be improved:

- remove old tests from `windows-build-test.yml`
- fix the drive letter issue in utils (maybe not an issue with the test itself)
- add some better scripting for analyzing dump files including remote kdb support
- add `!stacks` and `!memusage` to parsedump
- make some tests for previously fixed issues
- update the rest of the python tests to use `utils.py`
- update debug instructions, driver installation with the WDK deployment option fails a lot, maybe make it use zfsinstaller. we should get unmount / driver unload always working properly too

**Comment by sskras** [comment]
I would like to ask a beginner's question:

has the tests for Windows something to do with the test for *nix systems (like the common test-bootstrapping script or at least the common folder for different scripts or different, OS-specific subfolders?

**Reply by EchterAgo** [reply]
AFAICT they don't but I'm wondering if they should. I haven't really looked at the other tests yet.

**Comment by lundman** [comment]
OK so ultimately I want to run the `zfstest` suite of tests, it is in `ksh` script, thousand of files, and has a whole bunch of Unix assumptions, like `sudo`, `uid`, `groups` and all that.

The best way around that is to run `zfstest` in WSL shell, and have it execute `zfs` as `zfs.exe` in "Windows mode", to do things to the drives mounted/mapped to WSL.

We have a proof-of-concept of that somewhere, and it can run a few tests with some prep. One is to make the Windows drive "map" to Unix - it would be good if this could be automated.


Beyond that, we do need a few more tests, that are more Windows specific. Running `ifstest.exe` is a good idea, since we will need to make sure that passes to appease the Microsoft elders. And certainly no reason not to run btrfs' `test.exe` as well.
We also have a few tests done for things we have already come across. We should keep adding things here that are Windows specific.

But I do not think we should add ZFS specific tests, they should be part of `zfstests` to ensure compliance with the tests upstream has, and adds to.

**Reply by EchterAgo** [reply]
I got a self hosted github runner installed on a beefy machine now and will use that to try to do the WSL thing. Should be much easier to debug when I have direct access to the runner.

**Reply by EchterAgo** [reply]
The new runner is quite fast, much faster than the GitHub provided ones. It took some time to set it up, you basically have to build the image yourself from https://github.com/actions/runner-images, which only works on Azure. Azure didn't let me download the generated image to use on my own machine so I modified the packer configuration to build the image on one of my local `libvirtd`s.

After that I made a new image backed by the packer image, downloaded the runner and set it up. Note that the runner can not be set to run as `Network Service` or `Local Service`, it has to run as `Local System` or some other user account, otherwise the root certificate import fails because permission is denied.

I already managed to BSOD the virtual machine running the tests :\

**Reply by EchterAgo** [reply]
~~So it turns out the same runner runs all the tests in parallel. We could make this sequential to reduce failures, but I think our tests should be robust in this situation.~~ With this new self-hosted runner I can also still get information out after the machine crashes.

Edit: only the first 2 jobs are run on the same machine, the rest seems to run on github hosted runners. I might need more runners or some way to override `runs-on`

**Comment by EchterAgo** [comment]
I think we should also have a test that tests exercises lazy writes, flush on unmount and a check that everything has been written by importing the pool.

**Comment by sskras** [comment]
@andrebraga commented [30 minutes ago](https://github.com/openzfsonwindows/openzfs/issues/469#issuecomment-2834575357) in #469:

> for the time being having basic sanity checks ("as unprivileged user: can create file, can remove file just created; same with folders") would be great.

Sure, and maybe some of these ops are already being tested, eg:
https://github.com/openzfsonwindows/openzfs/blob/bce453c97aa3da40e322e76cc9c6c5a1485dbe0a/.github/workflows/windows-build-test.yml#L2507-L2514

I am just not yet sure about how does `winfstests.py` exactly work, eg:
https://github.com/openzfsonwindows/openzfs/blob/bce453c97aa3da40e322e76cc9c6c5a1485dbe0a/contrib/windows/tests/winfstests.py#L216-L235

What are all these file names like `.\t\base\08.t`, please ?
